File: dsa-problems/ProbabilisticDS/solutions/javascript/PDS-008-bottom-k-sampling.js
const readline = require("readline");

function jaccardEstimate(a, b) {
  let matches = 0;
  for (let i = 0; i < a.length; i++) {
    if (a[i] === b[i]) {
      matches++;
    }
  }
  return matches / a.length;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => {
  const parts = line.trim().split(/\s+/);
  for (const part of parts) {
    if (part !== "") data.push(part);
  }
});
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const k = parseInt(data[idx++], 10);
  const a = [];
  const b = [];
  for (let i = 0; i < k; i++) a.push(parseFloat(data[idx++]));
  for (let i = 0; i < k; i++) b.push(parseFloat(data[idx++]));
  console.log(jaccardEstimate(a, b).toFixed(6));
});

File: dsa-problems/ProbabilisticDS/solutions/python/PDS-008-bottom-k-sampling.py
import sys

def jaccard_estimate(a, b):
    matches = 0
    k = len(a)
    for i in range(k):
        if a[i] == b[i]:
            matches += 1
    return matches / k

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        k = int(next(iterator))
        a = []
        for _ in range(k):
            a.append(float(next(iterator)))
        b = []
        for _ in range(k):
            b.append(float(next(iterator)))
            
        print(f"{jaccard_estimate(a, b):.6f}")
    except StopIteration:
        pass

if __name__ == "__main__":
    main()
